modules/cassandrabackend/cassandrabackend.cc

```diff
@@ -0,0 +1,374 @@
+		cassandradbmanager::request_timeout = request_timeout;
+		cassandradbmanager::enable_load_balance_round_robin = enable_load_balance_round_robin;
+		cassandradbmanager::enable_token_aware_routing = enable_token_aware_routing;
+		cassandradbmanager::enable_latency_aware_routing = enable_latency_aware_routing;
+		cassandradbmanager::enable_tcp_nodelay = enable_tcp_nodelay;
+		cassandradbmanager::enable_tcp_keepalive = enable_tcp_keepalive;
+		if(::arg().mustDo("query-logging")) {
+			L << Logger::Info << "[CassandraBackend] Connecting to cassandra cluster " << endl;
+		}
+		cassandradbmanager::getInstance();
+		if(::arg().mustDo("query-logging")) {
+			L << Logger::Info << "[CassandraBackend] Connection to cassandra cluster successful " << endl;
+		}
+
+	 } catch (const ArgException &A) {
+		L << Logger::Error << "[CassandraBackend]" << " Fatal argument error: "<< A.reason << endl;
+		throw;
+	 } catch (...) {
+		throw;
+	 }
+
+  }
+
+  bool list(const DNSName& target, int domain_id, bool include_disabled=false)
+  {
+		try {
+			//Right now there is no disabled field in the domain field
+			cassandradbmanager *sc1 = cassandradbmanager::getInstance();
+			const char* query =
+					"SELECT domain FROM pdns.domain_id_domain_relation WHERE domain_id = ?";
+			if(::arg().mustDo("query-logging")) {
+				L << Logger::Info << "[CassandraBackend] SELECT domain FROM pdns.domain_id_domain_relation WHERE domain_id = " << domain_id << endl;
+			}
+			string param = sc1->executeAxfrQuery(query, domain_id);
+			if (param.size() > 0) {
+				this->queryType = "ANY";
+				fetchdata_multiple(param);
+				return true;
+			} else {
+				return false;
+			}
+		} catch (const std::exception &exc) {
+			L << Logger::Critical << "LIST exception " << exc.what() << endl;
+		} catch (...) {
+			L << Logger::Critical << "Unknown exception in LIST" << endl;
+		}
+		return false;
+	}
+
+
+  bool getSOA(const DNSName &name, SOAData &soadata, DNSPacket *p = 0) {
+		try {
+			domain = name.toStringNoDot();
+			if(::arg().mustDo("query-logging")) {
+			L << Logger::Info << "[CassandraBackend] Recieved getSOA " << domain << endl;
+			}
+			fetchdata();
+			vector<backendrecord>::const_iterator cii;
+			for(cii=backendRecords.begin(); cii!=backendRecords.end(); cii++) {
+			//for (int index = 0; index < totalSize; ++index) {
+				backendrecord backendRecord = (*cii);//backendRecords[index];
+				if (backendRecord.getType() == QType::SOA) {
+					if(::arg().mustDo("query-logging")) {
+						L << Logger::Info << "[CassandraBackend] SOA record found" << endl;
+					}
+					soadata.db = this;
+					//Data format :: domain_id nameserver	hostname    serial refresh retry expiry default_ttl
+					std::string data = backendRecord.getRecord();
+					boost::char_separator<char> delimiter("#");
+					boost::tokenizer<boost::char_separator<char> > tokens(data,
+							delimiter);
+					int rec_index = 0;
+					BOOST_FOREACH(std::string const& token, tokens) {
```

Comment:
Maybe instead of this just validate the size of the tokens vector and access with index directly?